include/swift/Runtime/Debug.h

```diff
@@ -134,6 +134,56 @@ void dumpStackTraceEntry(unsigned index, void *framePC,
 LLVM_ATTRIBUTE_NOINLINE
 void printCurrentBacktrace(unsigned framesToSkip = 1);
 
+/// Debugger breakpoint ABI. This structure is passed to the debugger (and needs
+/// to be stable) and describes extra information about a fatal error or a
+/// non-fatal warning, which should be logged as a runtime issue. Please keep
+/// all integer values pointer-sized.
+struct RuntimeErrorDetails {
+  // ABI version, needs to be "1" currently.
+  uintptr_t version;
+
+  // A short hyphenated string describing the type of the issue, e.g.
+  // "precondition-failed" or "exclusivity-violation".
+  const char *errorType;
+
+  // Description of the current thread's stack position.
+  const char *currentStackDescription;
+
+  // Address of some associated object (if there's any).
+  void *memoryAddress;
+
+  // A structure describing an extra thread (and its stack) that is related.
+  struct Thread {
+    const char *description;
+    uint64_t threadID;
+    uintptr_t numFrames;
+    void **frames;
+  };
+
+  // Number of extra threads (excluding the current thread) that are related,
+  // and the pointer to the array of extra threads.
+  uintptr_t numExtraThreads;
+  Thread *threads;
+
+  // Description of a related object (other than 'memoryAddress').
+  struct RelatedObject {
+    const char *description;
+    const char *filename;
+    uintptr_t line;
+    uintptr_t column;
```

Comment:
We could add in debug info for these deprecated `@objc` thunks, which points at the declaration. However, that won't capture the location at which we should insert the `@objc` in the source.

To be fully general w.r.t. notes with Fix-Its would require something like:

    struct RuntimeFixIt {
      const char *filename;
      uintptr_t startLine;             // text in [startLine:startColumn, endLine:endColumn) gets replaced with replacementText
      uintptr_t startColumn;
      uintptr_t endLine;
      uintptr_t endColumn;
      const char *replacementText;
  }

and then both `RuntimeErrorDetails` and `RelatedObject` would get:

    uintptr_t numFixIts;
    RuntimeFixIt *fixIts;

Note that having Fix-Its like these for `RuntimeErrorDetails` means SE-0160 wouldn't need `RelatedObjects`. That said, if we add the Fix-Its and rename `RelatedObject` to `Note` (or `RuntimeNote`, or whatever), we'll have mirrored the diagnostic information that both the Swift and Clang frontends have... which is defensible and likely to capture everything we want.